File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    bool dfs(int cur, vector<vector<int>>& adj, vector<int>& visited, vector<int>& path, stack<int>& stk) {
        visited[cur] = 1;
        path[cur] = 1;

        for (int num : adj[cur]) {
            if (!visited[num]) {
                if (dfs(num, adj, visited, path, stk)) return true; 
            }
            else if (path[num]) {
                return true; 
            }
        }

        path[cur] = 0;
        stk.push(cur); 
        return false;
    }

    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
        for (auto& p : prerequisites) {
            adj[p[1]].push_back(p[0]); 
        }

        vector<int> visited(numCourses, 0), path(numCourses, 0);
        stack<int> stk;

        for (int i = 0; i < numCourses; ++i) {
            if (!visited[i]) {
                if (dfs(i, adj, visited, path, stk)) {
                    return {}; 
                }
            }
        }

        vector<int> ans;
        while (!stk.empty()) {
            ans.push_back(stk.top());
            stk.pop();
        }
        return ans;
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public boolean dfs(int i,boolean[] visited,boolean[] path, ArrayList<ArrayList<Integer>> list,ArrayList<Integer> ans) {
        visited[i] = true;
        path[i] = true;
        for(int num : list.get(i)) {
            if(!visited[num]) {
                if(dfs(num,visited,path,list,ans)) return true;
            }
            else if(path[num]) return true;
        }
        ans.add(i);
        path[i] = false;
        return false;
    }
    public int[] findOrder(int n, int[][] nums) {
        ArrayList<ArrayList<Integer>> list = new ArrayList<>();
        for(int i=0; i<n; i++) {
            list.add(new ArrayList<>());
        }

        for(int arr[] : nums) {
            list.get(arr[1]).add(arr[0]);
        }

        boolean visited[] = new boolean[n];
        boolean path[] = new boolean[n];
        ArrayList<Integer> ans = new ArrayList<>();
        for(int i=0; i<n; i++) {
            if(!visited[i]) {
                if(dfs(i,visited,path,list,ans)) return new int[]{};
            }
        }
        Collections.reverse(ans);
        int x[] = new int[ans.size()];
        for(int i=0; i<list.size(); i++) {
            x[i] = ans.get(i);
        }
        return x;
        
    }
}